deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AutoPromoter.java

```diff
@@ -481,29 +470,75 @@ public Date getScheduledCheckDueTime(long start_date, String cronExpressionStrin
+     * @param envBean
+     * @param interval
+     * @param size
+     * @return
+     * @throws Exception
+     */
     List<BuildBean> getBuildCandidates(EnvironBean envBean, Interval interval, int size) throws Exception {
         // By default, buildName is the same as envName
         String buildName = envBean.getBuild_name();
         String scmBranch = envBean.getBranch();
-
-        return buildDAO.getAcceptedBuilds(buildName, scmBranch, interval, size);
+        List<BuildBean> taggedGoodBuilds = new ArrayList<BuildBean>();
+
+        List<BuildBean> availableBuilds = buildDAO.getAcceptedBuilds(buildName, scmBranch, interval, size);
+        if(!availableBuilds.isEmpty()) {
+            List<BuildTagBean> buildTagBeanList = buildTagsManager.getEffectiveTagsWithBuilds(availableBuilds);
+            for(BuildTagBean buildTagBean: buildTagBeanList) {
+                if(buildTagBean.getTag() != null && buildTagBean.getTag().getValue() == TagValue.BAD_BUILD) {
+                    // bad build,  do not include
+                    LOG.info("Env {} Build {} is tagged as BAD_BUILD, ignore", envBean.getEnv_id(), buildTagBean.getBuild());
+                } else {
+                    taggedGoodBuilds.add(buildTagBean.getBuild());
+                }
+            }
+        }
+        return taggedGoodBuilds;
     }
 
+    /**
+     * get a list of available deploys, and filter out the deploys with BAD_BUILD builds
+     * @param envId
+     * @param interval
+     * @param size
+     * @return
+     * @throws Exception
+     */
     List<DeployBean> getDeployCandidates(String envId, Interval interval, int size) throws Exception {
         LOG.info("Search Deploy candidates between {} and {} for environment {}",
             interval.getStart().toString(ISODateTimeFormat.dateTime()),
             interval.getEnd().toString(ISODateTimeFormat.dateTime()),
             envId);
-        return deployDAO.getAcceptedDeploys(envId, interval, size);
-    }
-
-    TagBean getBuildTag(DeployBean deployBean) throws Exception {
-        return getBuildTag(deployBean.getBuild_id());
-    }
-
-    TagBean getBuildTag(String buildId) throws Exception {
-        BuildBean buildBean = buildDAO.getById(buildId);
-        return buildTagsManager.getEffectiveBuildTag(buildBean);
+        List<DeployBean> taggedGoodDeploys = new ArrayList<DeployBean>();
+
+        List<DeployBean> availableDeploys = deployDAO.getAcceptedDeploys(envId, interval, size);
+        if(!availableDeploys.isEmpty()) {
+            List<String> availableBuildIds = new ArrayList<String>();
+            Map<String, DeployBean> buildId2DeployBean = new HashMap<String, DeployBean>();
+
+            for(DeployBean deployBean: availableDeploys) {
+                String buildId = deployBean.getBuild_id();
+                if(StringUtils.isNotEmpty(buildId)) {
+                    buildId2DeployBean.put(buildId, deployBean);
+                    availableBuildIds.add(deployBean.getBuild_id());
```

Comment:
What if we have multiple deploys with the same build? We may add duplicated build id here and some accept deploy won't be int he list. Maybe we should just create a build_id hashmap, then get tags and mark the bad. Then loop through the available deploys and only copy good build to another list.

Reply:
good point, just updated